File: MultilanguageSDK/Python/Samples/Morph/CameraModelSigleFrameDelayTest/CameraModelSigleFrameDelayTest.py
from pickle import FALSE, TRUE
import sys
currentPath =  sys.path[0]
pos = currentPath.find('Samples')
libpath = currentPath[:pos]
sys.path.append(libpath) #absolutely path

from API.ScepterDS_api import *
import time
import psutil

def get_current_pid():
    # 获取当前进程的PID
    current_pid = psutil.Process().pid
    return current_pid

def get_memory_usage(pid):
    # 获取指定PID的内存占用，单位为MB
    process = psutil.Process(pid)
    memory_info = process.memory_info()
    mem_usage = memory_info.rss / (1024 ** 2)  # rss表示已使用的物理内存
    return mem_usage

def get_cpu_usage(pid):
    # 获取指定PID的CPU占用百分比
    process = psutil.Process(pid)
    cpu_percent = process.cpu_percent(interval=0.5)  # 这里interval参数可选，用于统计间隔时间内的CPU使用率
    return cpu_percent


camera = ScepterTofCam()

def displayKeyconfiguration():
    ret, res = camera.scGetSDKVersion()
    if ret != 0:
        print("scGetSDKVersion failed status:", ret)
        return -1
    print("********scGetSDKVersion: ",res," ********")
    ret, res = camera.scGetFirmwareVersion()
    if ret != 0:
        print("scGetFirmwareVersion failed status:", ret)
        return -1
    print("********scGetFirmwareVersion: ",res," ********")
    ret, res = camera.scAIModuleGetEnabled()
    if ret != 0:
        print("scAIModuleGetEnabled failed status:", ret)
        return -1
    print("********scAIModuleGetEnabled: ",res," ********")
    if res == True:
        print("******** Config File ERROR, Please check the config file first ********")
        # ret = camera.scAIModuleSetEnabled(c_bool(False))
        # if ret != 0:
        #     print("scAIModuleGetEnabled failed status: ",ret)
        #     return -1

    ret, res = camera.scAIModuleGetInputFrameTypeEnabled(ScFrameType.SC_COLOR_FRAME)
    if ret != 0:
        print("scAIModuleGetInputFrameTypeEnabled SC_COLOR_FRAME failed status:", ret)
        return -1
    print("********scAIModuleGetInputFrameTypeEnabled SC_COLOR_FRAME : ",res," ********")

    ret, res = camera.scAIModuleGetInputFrameTypeEnabled(ScFrameType.SC_TRANSFORM_DEPTH_IMG_TO_COLOR_SENSOR_FRAME)
    if ret != 0:
        print("scAIModuleGetInputFrameTypeEnabled SC_TRANSFORM_DEPTH_IMG_TO_COLOR_SENSOR_FRAME failed status:", ret)
        return -1
    print("********scAIModuleGetInputFrameTypeEnabled SC_TRANSFORM_DEPTH_IMG_TO_COLOR_SENSOR_FRAME: ",res," ********")

    ret, res = camera.scAIModuleGetPreviewFrameTypeEnabled(ScFrameType.SC_COLOR_FRAME)
    if ret != 0:
        print("scAIModuleGetPreviewFrameTypeEnabled SC_COLOR_FRAME failed status:", ret)
        return -1
    print("********scAIModuleGetPreviewFrameTypeEnabled SC_COLOR_FRAME: ",res," ********")

    ret, res = camera.scAIModuleGetPreviewFrameTypeEnabled(ScFrameType.SC_IR_FRAME)
    if ret != 0:
        print("scAIModuleGetPreviewFrameTypeEnabled SC_IR_FRAME failed status:", ret)
        return -1
    print("********scAIModuleGetPreviewFrameTypeEnabled SC_IR_FRAME: ",res," ********")

    ret, res = camera.scAIModuleGetPreviewFrameTypeEnabled(ScFrameType.SC_DEPTH_FRAME)
    if ret != 0:
        print("scAIModuleGetPreviewFrameTypeEnabled SC_DEPTH_FRAME failed status:", ret)
        return -1
    print("********scAIModuleGetPreviewFrameTypeEnabled SC_DEPTH_FRAME: ",res," ********")

    ret, res = camera.scGetFrameRate()
    if ret != 0:
        print("scGetFrameRate failed status:", ret)
        return -1
    else:
        print("********scGetFrameRate : ", res, "********")


    res,pW,pH = camera.scGetColorResolution()
    if ret != 0:
        print("scGetColorResolution failed status:", ret)
        return -1
    else:
        print("********scGetColorResolution pW : ", pW, " pH: ", pH,"********")


print("---CameraModelSigleFrameDelayTest---")
camera_count = camera.scGetDeviceCount(3000)
print("Get device count:", camera_count)
if camera_count <= 0:
    print("scGetDeviceCount scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples.")
    exit()

device_info=ScDeviceInfo()

if camera_count > 0:
    ret,device_infolist=camera.scGetDeviceInfoList(camera_count)
    if ret==0:
        device_info = device_infolist[0]
    else:
        print('GetDeviceListInfo failed status:' + ret)
        exit()  
else: 
    print("there are no camera found")
    exit()

if  ScConnectStatus.SC_CONNECTABLE.value != device_info.status:
	print("connect status:",device_info.status)  
	print("The device state does not support connection.")
	exit()
else:
    print("productName: "+str(device_info.productName))
    print("serialNumber: "+str(device_info.serialNumber))
    print("ip: "+str(device_info.ip))
    print("connectStatus: "+str(device_info.status))

ret = camera.scOpenDeviceBySN(device_info.serialNumber)
if  ret != 0:
    print('OpenDevice failed status:' + str(ret))
    exit()

displayKeyconfiguration()
ret, frameRate = camera.scGetFrameRate()
if ret == 0:
    print("frameRate :", frameRate)
else:
    print("scGetFrameRate :", ret)
    exit()

ret = camera.scSetWorkMode(ScWorkMode.SC_SOFTWARE_TRIGGER_MODE)
if  ret != 0:  
    print("scSetWorkMode failed status:",ret)

ret = camera.scStartStream()
if  ret != 0:
    print("scStartStream failed status:"+ str(ret))
    exit()

curPath = os.getcwd()
filename = curPath + "/CameraModelSigleFrameDelayTest.csv"
try:
    file = open(filename, "w")
except IOError:
    print('csv file open failed')
    exit()
file.write("frameIndex,cpu,mem,TotalDelay,ExcludeDelayofExposure\n")

'''
//1.software trigger.
//2.ReadNextFrame.
//3.GetFrame acoording to Ready flag and Frametype.
//4.sleep 1000/frameRate (ms)
'''
number = int(input('Please input the number of tests:'))
for i in range(number):
    ret = camera.scSoftwareTriggerOnce()
    startTime = int(time.time()*1000)
    if  ret != 0:  
        print("scSoftwareTriggerOnce failed:",ret)

    ret, frameready = camera.scGetFrameReady(c_uint16(1200))
    if  ret != 0:
        print("scGetFrameReady failed status:",ret)
        continue       
    if  frameready.depth:      
        ret,frame = camera.scGetFrame(ScFrameType.SC_DEPTH_FRAME)
        endTime = int(time.time()*1000)
        if  ret == 0:
            if 0 == frame.frameIndex % 10:
                print("scGetFrame,status:", ret, "  ", "frameType:", frame.frameType, "  ", "frameIndex:", frame.frameIndex)
            exposureEndTime = frame.hardwaretimestamp
            frameIntervalTime = endTime - startTime
            frameIntervalNtpTime = endTime - exposureEndTime
            current_pid = get_current_pid()
            mem_usage = get_memory_usage(current_pid)
            cpu_usage = get_cpu_usage(current_pid)
            #print(f"当前进程PID: {current_pid}, 内存占用: {mem_usage:.2f} MB,CPU占用: {cpu_usage}%")
            if frameIntervalNtpTime > 2000:
                file.write("{0},{1},{2},{3},N/A\n".format(frame.frameIndex, cpu_usage, mem_usage,frameIntervalTime))
            else:
                file.write("{0},{1},{2},{3},{4}\n".format(frame.frameIndex, cpu_usage, mem_usage, frameIntervalTime, frameIntervalNtpTime))
    time.sleep(1/frameRate)
file.close()

ret = camera.scSetWorkMode(ScWorkMode.SC_ACTIVE_MODE)
if  ret != 0:  
    print("scSetWorkMode failed status:",ret)

ret = camera.scStopStream()
if  ret != 0:
    print("stopstream failed",ret)

ret = camera.scCloseDevice()     
if  ret != 0:
    print('scCloseDevice failed: ' + str(ret))
           

File: BaseSDK/Windows/Samples/Base/Morph/CameraModelSigleFrameDelayTest/CameraModelSigleFrameDelayTest.cpp
#include <iostream>
#ifdef _WIN32
#include <sys/timeb.h>
#else
#include <time.h>
#endif
#include <thread>
#include <fstream>
#include <string>
#include "Scepter_api.h"
#include "Scepter_Morph_api.h"
#include "get_cpu_vmem.h"
using namespace std;

uint64_t getCurrentTime()
{
#ifdef _WIN32
    timeb timeInfo;
    ftime(&timeInfo);
    return (timeInfo.time * 1000 + timeInfo.millitm);
#else
    timespec timeInfo;
    clock_gettime(CLOCK_REALTIME, &timeInfo);
    return (timeInfo.tv_sec * 1000 + timeInfo.tv_nsec/1000000);
#endif
}
int displayKeyconfiguration(const ScDeviceHandle deviceHandle)
{
	// key configuration display
	ScStatus status = SC_OTHERS;
	char version[64] = { 0 };
	status = scGetSDKVersion(version, 64);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetSDKVersion failed status:" << status << endl;
		return -1;
	}
	cout << "********scGetSDKVersion: " << version <<"********"<< endl;
	status = scGetFirmwareVersion(deviceHandle, version, 64);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetFirmwareVersion failed status:" << status << endl;
		return -1;
	}
	cout << "********scGetFirmwareVersion: " << version <<"********"<< endl;
	bool enable = false;
	status = scAIModuleGetEnabled(deviceHandle, &enable);
	if (status != ScStatus::SC_OK)
	{
		cout << "scAIModuleGetEnabled failed status:" << status << endl;
		return -1;
	}
	cout << "********scAIModuleGetEnabled: " << enable <<"********"<< endl;
	if (enable)
	{
		cout << "******** Config File ERROR, Please check the config file first ********"<< endl;
		// cout << "******** disable AI ********"<< endl;
		// status = scAIModuleSetEnabled(deviceHandle, false);
		// if (status != ScStatus::SC_OK)
		// {
		// 	cout << "scAIModuleGetEnabled failed status:" << status << endl;
		// 	return -1;
		// }
	}

	ScWorkMode pMode;
	status = scGetWorkMode(deviceHandle, &pMode);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetWorkMode failed status:" << status << endl;
		return -1;
	}
	cout << "********scGetWorkMode : " << pMode <<" ********"<< endl;
	ScAIModuleMode mode;
	status = scAIModuleGetWorkMode(deviceHandle, &mode);
	if (status != ScStatus::SC_OK)
	{
		cout << "scAIModuleGetWorkMode failed status:" << status << endl;
		return -1;
	}
	cout << "********scAIModuleGetWorkMode : " << mode <<" ********"<< endl;

	status = scAIModuleGetInputFrameTypeEnabled(deviceHandle, SC_COLOR_FRAME, &enable);
	if (status != ScStatus::SC_OK)
	{
		cout << "scAIModuleGetInputFrameTypeEnabled failed status:" << status << endl;
		return -1;
	}
	cout << "********scAIModuleGetInputFrameTypeEnabled SC_COLOR_FRAME: " << enable <<" ********"<< endl;
	status = scAIModuleGetInputFrameTypeEnabled(deviceHandle, SC_TRANSFORM_DEPTH_IMG_TO_COLOR_SENSOR_FRAME, &enable);
	if (status != ScStatus::SC_OK)
	{
		cout << "scAIModuleGetInputFrameTypeEnabled failed status:" << status << endl;
		return -1;
	}
	cout << "********scAIModuleGetInputFrameTypeEnabled SC_TRANSFORM_DEPTH_IMG_TO_COLOR_SENSOR_FRAME :" << enable <<" ********"<< endl;

	status = scAIModuleGetPreviewFrameTypeEnabled(deviceHandle, SC_COLOR_FRAME, &enable);
	if (status != ScStatus::SC_OK)
	{
		cout << "scAIModuleGetPreviewFrameTypeEnabled failed status:" << status << endl;
		return -1;
	}
	cout << "********scAIModuleGetPreviewFrameTypeEnabled SC_COLOR_FRAME: " << enable <<" ********"<< endl;
	status = scAIModuleGetPreviewFrameTypeEnabled(deviceHandle, SC_DEPTH_FRAME, &enable);
	if (status != ScStatus::SC_OK)
	{
		cout << "scAIModuleGetPreviewFrameTypeEnabled failed status:" << status << endl;
		return -1;
	}
	cout << "********scAIModuleGetPreviewFrameTypeEnabled SC_DEPTH_FRAME: " << enable <<" ********"<< endl;

	status = scAIModuleGetPreviewFrameTypeEnabled(deviceHandle, SC_IR_FRAME, &enable);
	if (status != ScStatus::SC_OK)
	{
		cout << "scAIModuleGetPreviewFrameTypeEnabled failed status:" << status << endl;
		return -1;
	}
	cout << "********scAIModuleGetPreviewFrameTypeEnabled SC_IR_FRAME: " << enable <<" ********"<< endl;
	int frameRate = 0;
	status = scGetFrameRate(deviceHandle, &frameRate);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetFrameRate failed status:" << status << endl;
		return -1;
	}
	else
	{
		cout << "********scGetFrameRate : " << frameRate << " ********" << endl;
	}
	int32_t pW;
	int32_t pH;
	status = scGetColorResolution(deviceHandle, &pW, &pH);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetColorResolution failed status:" << status << endl;
		return -1;
	}
	else
	{
		cout << "********scGetColorResolution pw: " << pW << " ph: " << pH << " ********" << endl;
	}
	ScTimeFilterParams pParams;
	scGetTimeFilterParams(deviceHandle, &pParams);
	cout << "********scGetTimeFilterParams: " << pParams.enable << " ********"<< endl;
}
int main()
{
	// Must use GUI tool import specific configuration first
	cout << "---CameraModelSigleFrameDelayTest---" << endl;
	ScStatus status = SC_OTHERS;

	//SDK Initialize
	status = scInitialize();
	if (status != ScStatus::SC_OK)
	{
		cout << "scInitialize failed status:" << status << endl;
		system("pause");
		return -1;
	}

	uint32_t deviceCount = 0;
	ScDeviceInfo* pDeviceListInfo = NULL;
	ScDeviceHandle deviceHandle = 0;
	//1.Search and notice the count of devices.
	//2.get infomation of the devices. 
	//3.open devices accroding to the info.
	status = scGetDeviceCount(&deviceCount, 3000);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetDeviceCount failed! make sure pointer valid or called scInitialize()" << endl;
		system("pause");
		return -1;
	}
	cout << "Get device count: " << deviceCount << endl;
	if (0 == deviceCount)
	{
		cout << "scGetDeviceCount scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples." << endl;
		system("pause");
		return -1;
	}

	pDeviceListInfo = new ScDeviceInfo[deviceCount];
	status = scGetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status == ScStatus::SC_OK)
	{
		if (SC_CONNECTABLE != pDeviceListInfo[0].status)
		{
			cout << "connect status: " << pDeviceListInfo[0].status << endl;
			cout << "The device state does not support connection." << endl;
			return -1;
		}
	}
	else
	{
		cout << "GetDeviceListInfo failed status:" << status << endl;
		return -1;
	}

	cout << "productName: " << pDeviceListInfo[0].productName << endl
	 	<< "serialNumber:" << pDeviceListInfo[0].serialNumber << endl
		<< "ip:" << pDeviceListInfo[0].ip << endl
		<< "connectStatus:" << pDeviceListInfo[0].status << endl;

	status = scOpenDeviceBySN(pDeviceListInfo[0].serialNumber, &deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "OpenDevice failed status:" << status << endl;
		return false;
	}
	// set soft work model
	status = scSetWorkMode(deviceHandle, SC_SOFTWARE_TRIGGER_MODE);
	if (status != ScStatus::SC_OK)
	{
		cout << "scSetWorkMode failed status:" << status << endl;
		return -1;
	}
	int frameRate = 0;
	status = scGetFrameRate(deviceHandle, &frameRate);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetFrameRate failed status:" << status << endl;
		return -1;
	}
	displayKeyconfiguration(deviceHandle);
	//Starts capturing the image stream
	status = scStartStream(deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scStartStream failed status:" << status << endl;
		return -1;
	}
	//Wait for the device to upload image data
	ScFrameReady FrameReady = { 0 };
	ScFrame depthFrame = { 0 };
	ofstream csvWriter;
	string fileName = "";
	csvWriter.open(fileName.append("CameraModelSigleFrameDelayTest.csv"));
	if (!csvWriter.is_open())
	{
		cout << "csv file open failed" << endl;
		return -1;
	}
	csvWriter << "frameIndex,cpu,mem,TotalDelay,ExcludeDelayofExposure" << endl;
	//1.software trigger.
	//2.ReadNextFrame.
	//3.GetFrame acoording to Ready flag and Frametype.
	//4.sleep 1000/frameRate (ms)
	uint64_t endTimestamp = 0;
	uint64_t startTimestamp = 0;
	uint64_t deviceTimestamp = 0;
	uint64_t frameInterval = 0;
	uint64_t frameIntervalNTP = 0;
	int current_pid = GetCurrentPid();
	float cpu = 0;
	float vmem = 0;
	cout << "Please input the number of tests:" << endl;
	uint32_t number = 0;
	cin >> number;
	for (int i = 0; i < number; i++)
	{
		//call the below api to trigger one frame, then the frame will be sent
		// if do not call this function, the frame will not be sent and the below call will return timeout fail
		status = scSoftwareTriggerOnce(deviceHandle);
		// ftime(&timeStart); // record the start timestamp
		startTimestamp = getCurrentTime();
		if (status != ScStatus::SC_OK)
		{
			cout << "scSoftwareTriggerOnce failed status:" << status << endl;
			continue;
		}
		//If no image is ready within 1200ms, the function will return ScRetGetFrameReadyTimeOut
		status = scGetFrameReady(deviceHandle, 1200, &FrameReady);
		if (status != ScStatus::SC_OK)
		{
			cout << "scGetFrameReady failed status:" << status << endl;
			continue;
		}
		if (1 == FrameReady.depth)
		{
			status = scGetFrame(deviceHandle, SC_DEPTH_FRAME, &depthFrame);
			if (depthFrame.pFrameData != NULL)
			{
				if (depthFrame.frameIndex % 10 == 0)
				{
					cout << "scGetFrame,status:" << status << "  "
						<< "frameType:" << depthFrame.frameType << "  "
						<< "frameIndex:" << depthFrame.frameIndex << endl;
				}
				//ftime(&timeEnd);// record the end timestamp
				endTimestamp = getCurrentTime();
				deviceTimestamp = depthFrame.deviceTimestamp;
				frameInterval = endTimestamp - startTimestamp;
				frameIntervalNTP = endTimestamp - deviceTimestamp;
				cpu = GetCpuUsageRatio(current_pid);
				vmem = GetMemoryUsage(current_pid);
				if (frameIntervalNTP > 2000)
				{
					csvWriter << depthFrame.frameIndex << "," << cpu << "," << vmem << "," << frameInterval <<  ",N/A" << endl;
				}
				else
				{
					csvWriter << depthFrame.frameIndex << "," << cpu << "," << vmem << "," << frameInterval << "," << frameIntervalNTP << endl;
				}
			}
		}
		this_thread::sleep_for(chrono::milliseconds(1000 / frameRate));
	}
	csvWriter.close();

	// //set slave false
	// status = scSetWorkMode(deviceHandle, SC_ACTIVE_MODE);
	// if (status != ScStatus::SC_OK)
	// {
	// 	cout << "scSetWorkMode failed status:" << status << endl;
	// 	return -1;
	// }

	//Stop capturing the image stream
	status = scStopStream(deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scStopStream failed status:" << status << endl;
		return -1;
	}

	//1.close device
	//2.SDK shutdown
	status = scCloseDevice(&deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scCloseDevice failed status:" << status << endl;
		return -1;
	}
	status = scShutdown();
	if (status != ScStatus::SC_OK)
	{
		cout << "scShutdown failed status:" << status << endl;
		return -1;
	}
	cout << "---end---" << endl;

	return 0;
}
